### PR TITLE
contrib: add meson build

### PR DESCRIPTION
**Pull request by neheb**
This is a fairly basic meson file that builds only the library. Useful for stuff like WrapDB.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

### CODE REVIEW

**Comment by piponazo** [comment]
Do you think it is needed to provide some documentation explaining how this file is supposed to be used? Or developers experienced with Meson (I am not) will manage to find out easily?

**Comment by neheb** [comment]
meson is dead simple so probably not. unfortunately i broke my hand after this PR so any changes will have to wait.

**Comment by piponazo** [comment]
> meson is dead simple so probably not. unfortunately i broke my hand after this PR so any changes will have to wait.

Sorry to hear that! Take care

**Comment by neheb** [comment]
@piponazo Updated. One hand typing is hard :)

@eli-schwartz One issue I've noticed with the VSCode plugin is that it's not loading these ../../ files.

**Comment by eli-schwartz** [comment]
> @eli-schwartz One issue I've noticed with the VSCode plugin is that it's not loading these ../../ files.

The build system isn't in the root of the project, it's in a subdirectory. The files that it isn't loading are files which are, as far as Meson knows, arbitrary files retrieved from the operating system -- not part of the project itself. So I'm not surprised that it mishandles this.



> This is a fairly basic meson file that builds only the library. Useful for stuff like WrapDB.

This will not work, since the WrapDB -- and in general, use as a subproject -- requires there to be a meson.build file in the root of the project.

If it's a hard requirement to isolate the contributed meson.build files away from the project root, then it's possible to do what the zstd project does, and provide a minimal stub meson.build in `contrib/meson/` and have that do a `subdir()` into the actual implementation files.

This will probably not fix VSCode, and it still cannot be used in the WrapDB directly, but the way zstd works is that the WrapDB provides its own minimal stub meson.build, installs it into the project root, and has that load the "actual implementation files". Since the only thing that needs to be duplicated in both places is the stub loader, that means the actual important bits are tracked in the project's own git repository and doesn't need to be duplicated.

Really though, that's sort of awkward. :/ Much simpler to simply place the meson.build in the project root, which also lets people use a wrapped subproject using git development versions of exiv2 for testing purposes.

**Comment by neheb** [comment]
sure, but the intention is not to replace cmake. It's not up to parity yet. From my experience, people like CMake better.

If no one has any objections, I'll move to the root directory.

**Comment by eli-schwartz** [comment]
> sure, but the intention is not to replace cmake. It's not up to parity yet. From my experience, people like CMake better.

Sure, no problem. I'm simply observing that even for a build system which is documented as "unofficial, secondary, useful for embedding in other projects perhaps -- but should not be used by default for building the project", it probably still makes sense to launch that from the root directory.

It's "just" two files, and the root directory doesn't seem to be focusing on "lean and mean" -- there's more than 10 text files in the root directory, two png images, a whole bunch of dotfiles for tool configuration, and *3* directories containing various aspects of testing. Two more won't really hurt, and users will enter through the README file anyway (and there be told to type `cmake`).

In general I suspect that most software projects shouldn't have a steep tax on adding more files to the project root, as long as each file has a clear and distinct purpose.

**Comment by kmilos** [comment]
> If no one has any objections, I'll move to the root directory.

Sure, as long as you include the disclaimer/note in the main README as well?

**Comment by piponazo** [comment]
Fine for me too to have the meson files in the root of the project. I will give it a try once I see them there ;)

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+curl_dep = dependency('libcurl', disabler: true, required: get_option('curl'))
+if curl_dep.found()
+  deps += curl_dep
+endif
+
+zlib_dep = dependency('zlib', disabler: true, required: get_option('png'))
+if zlib_dep.found()
+  deps += zlib_dep
+endif
```
Comment (eli-schwartz) [inline meson.build:61]: As a matter of curiosity, why do these need to be `disabler: true`?

It only takes effect if the `.found()` method returns false, but everywhere you seem to first check for that anyway.

Aside: if you do *not* use disablers, then it is safe to append them unconditionally to `deps +=`. Not-found dependencies are no-ops by default and possess no compiler flags or linkable libraries, so they just do nothing when added... (that's why you also need to explicitly check `.found()` and set `EXV_HAVE_*`).

**Review by eli-schwartz — COMMENTED** [review]
> READNE-meson

Slight typo in the name. :)

File: meson.build (hunk 1 above)
Comment (neheb) [reply meson.build:61]: It’s mostly a hack to get the later set calls to not work and have the cmake and meson config files be more sinmilar.

**Comment by neheb** [comment]
> > READNE-meson
> 
> Slight typo in the name. :)

Yeah I can’t type with one hand.

**Comment by neheb** [comment]
Added basic xmpsdk support.

File: meson.build (hunk 1 above)
Comment (eli-schwartz) [reply meson.build]: The later `set()` calls are "working" and produce `#undef`.

Comment (neheb) [reply meson.build]: the difference is

```
> /* #undef EXV_HAVE_LIBINTL_H */
```

vs

```
> #undef EXV_HAVE_LIBINTL_H
```

**Comment by neheb** [comment]
With some MinGW testing, I found some bugs.

**Comment by neheb** [comment]
```
../include/exiv2/xmp_exiv2.hpp:401:28: warning: 'Exiv2::Xmpdatum::operator=' redeclared inline; 'dllimport' attribute ignored [-Wignored-attributes]
inline Xmpdatum& Xmpdatum::operator=(bool value) {
```

No idea how to fix this. Any ideas? ping @kmilos

**Comment by neheb** [comment]
@piponazo I see a warning about a bad function cast: https://github.com/Exiv2/exiv2/blob/main/src/basicio.cpp#L370

Is there a point to this block? I would think std::filesystem is enough.

**Comment by kmilos** [comment]
@neheb Need to add `EXIV2API` to both redeclarations on lines 401 and 406 I think.

**Comment by neheb** [comment]
But then it doesn't compile.

**Comment by piponazo** [comment]
> @piponazo I see a warning about a bad function cast: https://github.com/Exiv2/exiv2/blob/main/src/basicio.cpp#L370
> 
> Is there a point to this block? I would think std::filesystem is enough.

I think you are right. Note that this piece of code is very old and `std::filesystem` was not available at the moment. Probably the author of that code just wanted to use the native way to replace a file. Nowadays we should strive to use standard C++ whenever possible.

**Comment by kmilos** [comment]
> Nowadays we should strive to use standard C++ whenever possible.

I find this somehow at odds with trying to support older compilers.

**Comment by piponazo** [comment]
> > Nowadays we should strive to use standard C++ whenever possible.
> 
> I find this somehow at odds with trying to support older compilers.

In my opinion these are 2 different things:
* In this case, we could replace some piece of native code with standard C++ code. This would result in less code which would be potentially simpler to maintain and potentially support more platforms/compilers which we are not considering at the moment.
* The other case in discussion is about supporting some not so old gcc & clang versions which partial support of C++17. The Filesystem library was not fully added until more recent versions of those compilers, but an experimental one is available. In practice the implementation is very similar and we just need to change some header inclusions + linking flags depending on the version used of these compilers. The code would still be standard C++.

**Comment by kmilos** [comment]
Sure, it's doable, but IMHO it's completely legitimate to say we require C++17 and GCC 8 on main/next version in order to have a clean code base and reduce maintenance burden.

**Comment by piponazo** [comment]
> Sure, it's doable, but IMHO it's completely legitimate to say we require C++17 and GCC 8 on main/next version in order to have a clean code base and reduce maintenance burden.

That´s also true, we just need to agree on something 😉 . In #2136 the problem was with debian-10 (having gcc 8.3). In the newer stable debian-11 gcc version is 10.2 and it should not be a problem. If there are not major distros were this would be a problem, I would also go for simple code and clearly stating in the README that we require a C++17 compiler with non experimental support of the Filesystem library. @neheb  what do you think? Should we abandom #2484 and just add some info in the README?

**Comment by kmilos** [comment]
Gcc 8 (Debian 10, Ubuntu 18.4) should have it, that bug was just about missing linker lib: https://github.com/Exiv2/exiv2/issues/2136#issuecomment-1063855794

**Comment by neheb** [comment]
I see no problem with https://github.com/Exiv2/exiv2/pull/2484

**Comment by neheb** [comment]
```
"cl" "-Iexiv2-1.dll.p" "-I." "-I.." "-I..\include\exiv2" "-I..\xmpsdk\include" "-I..\subprojects\expat-2.5.0\lib" "/MDd" "/nologo" "/showIncludes" "/utf-8" "/Zc:__cplusplus" "/W2" "/EHsc" "/std:c++17" "/permissive-" "/Od" "/Zi" "-DEXIV2API=__declspec(dllexport)" "/Fdexiv2-1.dll.p\src_basicio.cpp.pdb" /Foexiv2-1.dll.p/src_basicio.cpp.obj "/c" ../src/basicio.cpp
../src/basicio.cpp(663): error C2589: '(': illegal token on right side of '::'
../src/basicio.cpp(663): error C2062: type 'unknown-type' unexpected
../src/basicio.cpp(663): error C2059: syntax error: ')'
../src/basicio.cpp(831): error C2589: '(': illegal token on right side of '::'
../src/basicio.cpp(831): error C2062: type 'unknown-type' unexpected
../src/basicio.cpp(831): error C2059: syntax error: ')'
../src/basicio.cpp(832): error C2589: '(': illegal token on right side of '::'
../src/basicio.cpp(832): error C2062: type 'unknown-type' unexpected
../src/basicio.cpp(832): error C2059: syntax error: ')'
../src/basicio.cpp(1080): error C2589: '(': illegal token on right side of '::'
../src/basicio.cpp(1080): error C2062: type 'unknown-type' unexpected
../src/basicio.cpp(1080): error C2059: syntax error: ')'
../src/basicio.cpp(1110): warning C4003: not enough arguments for function-like macro invocation 'max'
../src/basicio.cpp(1110): error C2589: '(': illegal token on right side of '::'
../src/basicio.cpp(1110): error C2144: syntax error: 'unknown-type' should be preceded by ')'
../src/basicio.cpp(1110): error C2660: 'Exiv2::RemoteIo::Impl::getDataByRange': function does not take 0 arguments
../src/basicio.cpp(1032): note: see declaration of 'Exiv2::RemoteIo::Impl::getDataByRange'
../src/basicio.cpp(1110): note: while trying to match the argument list '()'
../src/basicio.cpp(1110): error C2144: syntax error: 'unknown-type' should be preceded by ';'
../src/basicio.cpp(1110): error C2059: syntax error: ')'
../src/basicio.cpp(1110): error C2059: syntax error: ')'
../src/basicio.cpp(1110): warning C4003: not enough arguments for function-like macro invocation 'max'
../src/basicio.cpp(1110): error C2589: '(': illegal token on right side of '::'
../src/basicio.cpp(1118): error C2589: '(': illegal token on right side of '::'
../src/basicio.cpp(1118): error C2062: type 'unknown-type' unexpected
../src/basicio.cpp(1118): error C2059: syntax error: ')'
../src/basicio.cpp(1243): error C2589: '(': illegal token on right side of '::'
../src/basicio.cpp(1243): error C2062: type 'unknown-type' unexpected
../src/basicio.cpp(1243): error C2059: syntax error: ')'
../src/basicio.cpp(1261): error C2589: '(': illegal token on right side of '::'
../src/basicio.cpp(1261): error C2062: type 'unknown-type' unexpected
../src/basicio.cpp(1261): error C2059: syntax error: ')'
../src/basicio.cpp(1457): warning C4003: not enough arguments for function-like macro invocation 'max'
../src/basicio.cpp(1457): warning C4003: not enough arguments for function-like macro invocation 'max'
../src/basicio.cpp(1457): error C2589: '(': illegal token on right side of '::'
../src/basicio.cpp(1457): error C2062: type 'unknown-type' unexpected
../src/basicio.cpp(1457): error C2589: '(': illegal token on right side of '::'
../src/basicio.cpp(1457): error C2059: syntax error: ')'
../src/basicio.cpp(1457): error C2143: syntax error: missing ';' before '{'
[13/94] Compiling C++ object exiv2-1.dll.p/src_convert.cpp.obj
[14/94] Compiling C++ object exiv2-1.dll.p/src_datasets.cpp.obj
ninja: build stopped: subcommand failed.
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: C:\ProgramData\Chocolatey\bin\ninja.EXE -C D:/a/exiv2/exiv2/build
Error: Process completed with exit code 1.
```

Seems I need a hack to undefine max,

File: .github/workflows/meson.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
+name: meson checks
```
Comment (piponazo) [inline .github/workflows/meson.yml:1]: [suggestion] Could you follow the same naming schema we already have for the pipelines? For example, add the suffix `on_PR_` if you are thinking about running this pipeline on PRs

**Review by piponazo — DISMISSED** [review]

File: meson.build [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+  cargs += '-DEXIV2API=__declspec(dllexport)'
+elif cpp.has_function_attribute('visibility')
+  cargs += '-DEXIV2API=__attribute__((visibility("default")))'
+else
+  cargs += '-DEXIV2API='
+endif
+
+if host_machine.system() == 'windows'
+  if cpp.get_argument_syntax() == 'gcc'
+    add_project_arguments('-D__USE_MINGW_ANSI_STDIO', '-D__MINGW_USE_VC2005_COMPAT', language: 'cpp')
```
Comment (kmilos) [inline meson.build:21]: Any particular reason for controlling these directly? IMHO they should be left as determined by the MSYS2/MinGW-w64 toolchain.

Comment (kmilos) [reply meson.build:21]: Another platform we should probably stop supporting on main ;) Fine though...

Comment (neheb) [reply meson.build:21]: MSVC32bit since 2005 does 64-bit time_t btw.

File: .github/workflows/meson.yml (hunk 2 above)
Comment (neheb) [reply .github/workflows/meson.yml]: Done

Comment (piponazo) [reply .github/workflows/meson.yml]: Actually, I was referring to the filename 😅 . Apart from that the rest of the PR looks great and I tried to use meson to configure and compile the project and it worked like a charm 👏

**Review by piponazo — DISMISSED** [review]

File: .github/workflows/meson.yml (hunk 2 above)
Comment (neheb) [reply .github/workflows/meson.yml]: Renamed

**Review by piponazo — APPROVED** [review]

**Pull request merged**